docs: add third-party storage docs (#728)
harvester/docs: docs/advanced/csidriver.md

  <link rel="canonical" href="https://docs.harvesterhci.io/v1.4/advanced/csidriver"/>
</head>

_Available as of v1.5.0_

Harvester now supports provisioning of root volumes and data volumes using external [Container Storage Interface (CSI)](https://kubernetes-csi.github.io/docs/introduction.html) drivers. This enhancement allows you to select drivers that meet specific requirements, such as performance optimization or seamless integration with existing internal storage solutions.

:::important

> The matrix below shows the supported capabilities of the validated CSI drivers.

Reviewer: Does it make sense to include a href here to the SUSE certified storage for virtualization? E.g., _"For more information on SUSE certified 3rd-party storage, see https://www.suse.com/product-certification/suse-certified/virtualization-certification/."_

Author: Thanks @ihcsim,
Actually, I did not know this page. Do we already have the certified storage?

</head>

_Available as of v1.5.0_

Harvester now supports provisioning of root volumes and data volumes using external [Container Storage Interface (CSI)](https://kubernetes-csi.github.io/docs/introduction.html) drivers. This enhancement allows you to select drivers that meet specific requirements, such as performance optimization or seamless integration with existing internal storage solutions.

:::important

The matrix below shows the supported capabilities of the validated CSI drivers.

| Storage Solution | VM Image | VM Rook Disk | VM Data Disk | Volume Export To VM Image | VM Template Generator | VM Live Migration | VM Snapshot | VM Backup |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Longhorn v2 Data Engine | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; |&#10006; | &#10006; |
| Rook (RBD) | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; |
| LVM | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; | &#10004; | &#10006; |
| NFS | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; | &#10006; |

> For more information on SUSE certified 3rd-party storage. See [SUSE Certified Storage for Virtualization](https://www.suse.com/product-certification/suse-certified/virtualization-certification/).

Reviewer 2: This sentence is weird and looks like to be incomplete.

:::

## Prerequisites

> For the Harvester functions to work well, the third-party CSI driver should support the following capabilities:
> - Volume expansion, for online resizing of the volume
> - Volume snapshot, for taking snapshots of the volume/VM
> - Volume clone, for cloning the volume/VM
> - Support Read-Write-Many (RWX), for [Live Migration](../vm/live-migration.md)

Reviewer 3: ```suggestion
To enable Harvester to function well, use CSI drivers that support the following capabilities:
- Volume expansion (online resizing)
- Snapshot creation (volume and virtual machine snapshots)
- Cloning (volume and virtual machine clones)
- Usage of Read-Write-Many (RWX) volumes for [Live Migration](../vm/live-migration.md)
```

Reviewer 3: This suggestion was missed.

:::important

The matrix below shows the supported capabilities of the validated CSI drivers.

| Storage Solution | VM Image | VM Rook Disk | VM Data Disk | Volume Export To VM Image | VM Template Generator | VM Live Migration | VM Snapshot | VM Backup |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Longhorn v2 Data Engine | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; |&#10006; | &#10006; |
| Rook (RBD) | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; |
| LVM | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; | &#10004; | &#10006; |
| NFS | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; | &#10006; |

Also, here is the validated CSI driver and its corresponding storage solution:
- Longhorn v2 Data Engine: driver.longhorn.io
- Rook (RBD): rook-ceph.rbd.csi.ceph.com
- LVM: lvm.driver.harvesterhci.io
- NFS: nfs.csi.k8s.io

> For more information on SUSE certified 3rd-party storage, see [SUSE Certified Storage for Virtualization](https://www.suse.com/product-certification/suse-certified/virtualization-certification/).

Author: yeah, I thought we could remove this link at this moment.

![create-fs-volume](/img/v1.5/advanced/create-fs-volume.png)

## Advanced Topics

### Storage Profiles

You can now use the CDI API to create custom [storage profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) that simplify definition of data volumes. Storage profiles allow multiple data volumes to share the same provisioner settings.

The following is an example of an LVM storage profile:

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: StorageProfile
metadata:
  name: lvm-node-1-striped
spec:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
status:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
  cloneStrategy: snapshot
  dataImportCronSourceFormat: pvc
  provisioner: lvm.driver.harvesterhci.io
  snapshotClass: lvm-snapshot
  storageClass: lvm-node-1-striped
```

For more information, see [Storage Profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) in the CDI documentation.

You can define the above fields to override the default configuration showing on the status.

### Limitations

> - Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 

Reviewer: I think we should link this to the "VM Backup Compatibility" added in https://github.com/harvester/docs/pull/731, to provide user with more context. E.g., _"For more information, see [VM Backup Compatibility](#vm-backup-compatibility)."_

![various-volumes-for-vm-created](/img/v1.5/advanced/various-volumes-for-vm-created.png)

### Volume Creation

You can create volumes in your external storage solution.

When [creating a volume](../volume/create-volume.md) using the Harvester UI (**Volumes > Create**), you must perform the following actions:

- **Storage Class**: Select the target StorageClass, e.g. **nfs-csi**.
- **Volume Mode**: Select the corresponding volume mode, e.g. **Filesystem** for **nfs-csi**.

![create-fs-volume](/img/v1.5/advanced/create-fs-volume.png)

## Advanced/Internal Topics

### Top level architecture

By introducing the [Containerized-Data Importer (CDI)](https://kubevirt.io/user-guide/storage/containerized_data_importer/) into Harvester, Harvester can abstract the VM Image and VM Root Disk path for generic third-party storage solutions. The CDI will help to import the VM Image and VM Root Disk. So, the third-party storage solution can be used as the VM Image and VM Root Disk.

More detailed can be found in the HEP [Third Party Storage Support](https://github.com/harvester/harvester/blob/master/enhancements/20250203-third-party-storage-support.md).

> ### Notable Configuration Options

Reviewer 3: This heading is not necessary. Please remove it.

> #### Storage Profile
> 
> The CRD **StorageProfile** is the important concept for the Containerized-Data Importer (CDI). The **StorageProfile** can be used to define individual configuration for different storage providers. The following is the example for the **StorageProfile** for the LVM storage provider. We can use it to know each fields' meaning.

Reviewer 3: ```suggestion
### Storage Profiles

You can now use the CDI API to create custom [storage profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) that simplify definition of data volumes. Storage profiles allow multiple data volumes to share the same provisioner settings.

The following is an example of an LVM storage profile:
```

Reviewer 3: I made a few changes to Ivan's suggested wording and added the link to the CDI doc.

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: StorageProfile
metadata:
  name: lvm-node-1-striped
spec:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
status:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
  cloneStrategy: snapshot
  dataImportCronSourceFormat: pvc
  provisioner: lvm.driver.harvesterhci.io
  snapshotClass: lvm-snapshot
  storageClass: lvm-node-1-striped
```

> - claimPropertySets: The claimPropertySets field is used to define the accessModes and volumeMode for the StorageProfile.
> - snapshotClass: The snapshotClass field is used to define the VolumeSnapshotClass for the StorageProfile.
> - cloneStrategy: The cloneStrategy field is used to define the clone strategy for the StorageProfile.
> 
> The cloneStrategy is associated with the [efficient clone](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/efficient-cloning.md).
> 
> Efficient clone could be classified into three categories:
> - *CSI volume cloning* - Leverage the clone capability of the storage provider. The corresponding cloneStrategy is **csi-clone**.
> - *Smart Cloning* (snapshot-based cloning) - Leverage the snaphsot capability of the storage provider. The corresponding cloneStrategy is **snapshot**.
> - *host-assisted cloning* - Depends on directly data copy from source pod (volume) to target pod (volume). The corresponding cloneStrategy is **copy**.

Reviewer 3: ```suggestion
For more information, see [Storage Profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) in the CDI documentation.
```

Reviewer 3: The draft field descriptions do not provide a lot of information. Also, @ihcsim [suggested](https://github.com/harvester/docs/pull/728#issuecomment-2749602168) removing the clone strategies part.

You can define the above fields to override the default configuration showing on the status.

### Limitations

- Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 

> - A known issue in the CDI prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.

Reviewer 4: Suggest adding the upstream issue for reference.

Author: Could we use limitation instead of issue?
I will create the upstream issue, but that might need some discussion.

Reviewer 4: Sounds good.

Before you can make use of Harvester's **Backup & Snapshot** features, you need to set up some essential configurations through the Harvester [csi-driver-config](../advanced/settings.md#csi-driver-config) setting. Follow these steps to make these configurations:

:::note

Backup currently only supports the Longhorn v1 Data Engine. If you are using other storage providers, you can skip the **Backup VolumeSnapshot Class Name** configuration.

:::

1. Login to the Harvester UI, then navigate to **Advanced** > **Settings**.
1. Find and select **csi-driver-config**, and then select **⋮** > **Edit Setting** to access the configuration options.
1. Set the **Provisioner** to the third-party CSI driver in the settings.
1. Next, Configure the **Volume Snapshot Class Name**. This setting points to the name of the `VolumeSnapshotClass` used for creating volume snapshots or VM snapshots.

![csi-driver-config-external](/img/v1.5/advanced/csi-driver-config-external.png)

## Use the CSI driver

> After we have installed the CSI driver and configured the Harvester Cluster, we can now use the third-party storage solution in Harvester on serval scenarios.

Reviewer: ```suggestion
After we have installed the CSI driver and configured the Harvester Cluster, we can now use the third-party storage solution in Harvester on several scenarios.
```

### VM Image

Now, we can use the third-party storage solution as the VM image storage.

> 1. GoTo **Image** and Click **Create**

Reviewer: ```suggestion
1. Go to **Image** and Click **Create**
```

1. Config **URL** or **File** for the image
1. Config *StorageClass* on the **Storage** page. Select the StorageClass (e.g., *nfs-csi*) for your third-party storage solution.
1. Click **Create** to create the VM image.

![create-image-with-nfs-csi](/img/v1.5/advanced/create-image-with-nfs-csi.png)

After the image is created, you can see the image is created with the third-party storage solution.
![created-image-with-nfs-csi](/img/v1.5/advanced/created-image-with-nfs-csi.png) 

### VirtualMachine

Now, we can use the third-party storage solution as the VM root volume and data volume.

> 1. GoTo **Virtual Machine** and Click **Create**
> 1. GoTo **Volumes** and select the VM image created with the third-party storage solution.

Reviewer: ```suggestion
1. Go to **Virtual Machine** and click **Create**
1. Go to **Volumes** and select the VM image created with the third-party storage solution.
```

1. Click **Add Volume** to add the data volume with the third-party storage solution.
1. Click **Create** to create the VM.

The VM root volume is created with NFS, and the data volume is created with Longhorn v2 DataEngine.

![various-volumes-for-vm-creating](/img/v1.5/advanced/various-volumes-for-vm-creating.png)

![various-volumes-for-vm-created](/img/v1.5/advanced/various-volumes-for-vm-created.png)

### Volume

Since we support third-party storage solutions, you can create the volume and select the corresponding volume mode.

> 1. GoTo **Volumes** and Click **Create**

Reviewer: ```suggestion
1. Go to **Volumes** and click **Create**
```

1. Select the **StorageClass** for the third-party storage solution.
1. Select the *nfs-csi* StorageClass and you should also select the *Filesystem* for the volume mode.
1. Click **Create** to create the volume.

![create-fs-volume](/img/v1.5/advanced/create-fs-volume.png)

## Advanced/Internal Topics

### Top level architecture

> By introducing the Containerized-Data Importer (CDI) into Harvester, Harvester can abstract the VM Image and VM Root Disk path for generic third-party storage solutions. The CDI will help to import the VM Image and VM Root Disk. So, the third-party storage solution can be used as the VM Image and VM Root Disk.

Reviewer: ```suggestion
By introducing the [Containerized-Data Importer (CDI)](https://kubevirt.io/user-guide/storage/containerized_data_importer/) into Harvester, Harvester can abstract the VM Image and VM Root Disk path for generic third-party storage solutions. The CDI will help to import the VM Image and VM Root Disk. So the third-party storage solution can be used as the VM Image and VM Root Disk.
```

More detailed can be found in the HEP [Third Party Storage Support](https://github.com/harvester/harvester/blob/master/enhancements/20250203-third-party-storage-support.md).

### Notable Configuration Options

#### Storage Profile

> The CRD **StorageProfile** is the import concept for the Containerized-Data Importer (CDI). The **StorageProfile** can be used to define individual configuration for different storage providers. The following is the example for the **StorageProfile** for the LVM storage provider. We can use it to know each fields' meaning.

Reviewer: > import concept

What does this mean? Or did you mean "important concept"?

- cloneStrategy: The cloneStrategy field is used to define the clone strategy for the StorageProfile.

The cloneStrategy is associated with the [efficient clone](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/efficient-cloning.md).

Efficient clone could be classified into three categories:
- *CSI volume cloning* - Leverage the clone capability of the storage provider. The corresponding cloneStrategy is **csi-clone**.
- *Smart Cloning* (snapshot-based cloning) - Leverage the snaphsot capability of the storage provider. The corresponding cloneStrategy is **snapshot**.
- *host-assisted cloning* - Depends on directly data copy from source pod (volume) to target pod (volume). The corresponding cloneStrategy is **copy**.

You can define the above fields to override the default configuration showing on the status.

### Limitatioin

We still have the following limitations for current third-party storage support.

#### Backup is not supported

Currently, the backup feature only supports the Longhorn v1 Data Engine. If the VM contains any volume from the third-party storage provider, the backup feature will be blocked.

> #### Cannot convert the attached PVC to the VM Image

Author: updated

Before you can make use of Harvester's **Backup & Snapshot** features, you need to set up some essential configurations through the Harvester [csi-driver-config](../advanced/settings.md#csi-driver-config) setting. Follow these steps to make these configurations:

:::note

Backup currently only works with the Longhorn v1 Data Engine. If you are using other storage providers, you can skip the **Backup VolumeSnapshot Class Name** configuration.

> For more information, see [VM Backup Compatibility](../../versioned_docs/version-v1.4/advanced/csidriver.md#vm-backup-compatibility).

Reviewer 2: I think this link is wrong. It should point to the latest version.
```suggestion
For more information, see [VM Backup Compatibility](./csidriver.md#vm-backup-compatibility).
```

The following is an example of an LVM storage profile:

```yaml
apiVersion: cdi.kubevirt.io/v1beta1
kind: StorageProfile
metadata:
  name: lvm-node-1-striped
spec:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
status:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
  cloneStrategy: snapshot
  dataImportCronSourceFormat: pvc
  provisioner: lvm.driver.harvesterhci.io
  snapshotClass: lvm-snapshot
  storageClass: lvm-node-1-striped
```

For more information, see [Storage Profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) in the CDI documentation.

You can define the above fields to override the default configuration showing on the status.

### Limitations

- Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 

> - There is a limitation in the CDI prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.

Reviewer 2: ```suggestion
- There is a limitation in the CDI which prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.
```

kind: StorageProfile
metadata:
  name: lvm-node-1-striped
spec:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
status:
  claimPropertySets:
  - accessModes:
    - ReadWriteOnce
    volumeMode: Block
  cloneStrategy: snapshot
  dataImportCronSourceFormat: pvc
  provisioner: lvm.driver.harvesterhci.io
  snapshotClass: lvm-snapshot
  storageClass: lvm-node-1-striped
```

For more information, see [Storage Profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) in the CDI documentation.

You can define the above fields to override the default configuration showing on the status.

### Limitations

- Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 

- There is a limitation in the CDI which prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.

![convert-pvc-to-image-stuck](/img/v1.5/advanced/convert-pvc-to-image-stuck.png)

> ## Appendix

Author: Sounds good! Thanks for the explanation!

> ### How to deploy the NFS CSI driver

Reviewer: Why do we need to include this in our docs? It feels a bit out of place. If any of the upstream csi nfs instructions changed, our documentation will become stale. Also, the provisioner at https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner seems very old. If we are going to recommend it in our docs, we need to make sure their images are safe and updated.

Alternately, what do you of converting this content into a blog post or recording or kubecon talk proposal where we showcase how to use nfs csi driver and storage profile in 1.5? I can help with this (especially, kubecon proposal :wink:), if needed.

Reviewer: If we really want to include this in our docs, maybe consider just add a href to https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/deploy/example/README.md and call it done?

Reviewer: yeah, i prefer not to include it in our docs. many of its closed issues indicated that the project is no longer maintained. any reasons why we don't want to link to the csi nfs official docs at https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/deploy/example/README.md?

Author: OK, let's remove!

### How to deploy the NFS CSI driver

:::note

To deploy the NFS CSI driver, you should already have the NFS server installed and running.

:::

> - Install with csi-driver-nfs helm chart

Reviewer 3: ```suggestion
1. Install the driver using the `csi-driver-nfs` Helm chart.
```

  ```
  $ helm repo add csi-driver-nfs https://raw.githubusercontent.com/kubernetes-csi/csi-driver-nfs/master/charts
  $ helm install csi-driver-nfs csi-driver-nfs/csi-driver-nfs --namespace kube-system --version v4.10.0
  ```

> - create the StorageClass for NFS

Reviewer 3: ```suggestion
1. Create the StorageClass for NFS.
```

  ```yaml
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    name: nfs-csi
  provisioner: nfs.csi.k8s.io
  parameters:
    server: <your-nfs-server-ip>
    share: <your-nfs-share>
    # csi.storage.k8s.io/provisioner-secret is only needed for providing mountOptions in DeleteVolume
    # csi.storage.k8s.io/provisioner-secret-name: "mount-options"
    # csi.storage.k8s.io/provisioner-secret-namespace: "default"
  reclaimPolicy: Delete
  volumeBindingMode: Immediate
  allowVolumeExpansion: true
  mountOptions:
    - nfsvers=4.2
  ```

> After that, you can create VM Image/Root Volume/Data Volume with the StorageClass `nfs-csi`.

Reviewer 3: ```suggestion
  Once created, you can use the StorageClass to create virtual machine images, root volumes, and data volumes.
```

> ### How to deploy the AllInOne NFS Service + CSI driver

Reviewer 3: ```suggestion
### Deploy the Combined NFS Service and CSI Driver
```

Reviewer 3: There are no search results for "AllInOne NFS Service" so I assume this is not an official term.

Author: Yeah, you were correct!
Updated!

> - clone the AllInOne repo https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner

Reviewer 3: ```suggestion
1. Clone the [nfs-ganesha-server-and-external-provisioner repository](https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner).
```

  ```
  $ git clone https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner 
  ```

> - change the directory to `nfs-ganesha-server-and-external-provisioner/deploy/kubernetes` and modify the deployment.yaml file as follows:

Reviewer 3: ```suggestion
1. Change the directory to `nfs-ganesha-server-and-external-provisioner/deploy/kubernetes` and modify the `deployment.yaml` file as follows:
```

  ```
  spec:
    ...
    template:
      ...
      spec:
        ...
        nodeSelector:
          kubernetes.io/hostname: <your-nfs-export node>
        volumes:
          - name: export-volume
            hostPath:
              path: <your-nfs-export-path>
  ```

> - apply the deployment.yaml file

Reviewer 3: ```suggestion
1. Apply the `deployment.yaml` file.
```

  ```
  $ kubectl apply -f deployment.yaml
  ```

> - apply the RBAC

Reviewer 3: ```suggestion
1. Apply the RBAC configuration.
```

  ```
  $ kubectl apply -f rbac.yaml
  ```

> - create the StorageClass for NFS

Reviewer 3: ```suggestion
1. Create the StorageClass for NFS.
```

  ```
  $ kubectl apply -f class.yaml
  ```

> - The above StorageClass named example-nfs, so we need to update the corresponding StorageProfile CRD as below

Reviewer 3: ```suggestion
1. Update the StorageProfile CRD that corresponds to the StorageClass you created.

  In the following example, the StorageClass is `example-nfs`.
```

  ```yaml
  apiVersion: cdi.kubevirt.io/v1beta1
  kind: StorageProfile
  metadata:
    ...
    name: example-nfs
  spec:
    claimPropertySets:
    - accessModes:
      - ReadWriteMany
      volumeMode: Filesystem
  ```

> After that, you can create VM Image/Root Volume/Data Volume with the StorageClass `example-nfs`.

Reviewer 3: ```suggestion
  Once created, you can use the StorageClass to create virtual machine images, root volumes, and data volumes.
```

  ```yaml
  apiVersion: storage.k8s.io/v1
  kind: StorageClass
  metadata:
    name: nfs-csi
  provisioner: nfs.csi.k8s.io
  parameters:
    server: <your-nfs-server-ip>
    share: <your-nfs-share>
    # csi.storage.k8s.io/provisioner-secret is only needed for providing mountOptions in DeleteVolume
    # csi.storage.k8s.io/provisioner-secret-name: "mount-options"
    # csi.storage.k8s.io/provisioner-secret-namespace: "default"
  reclaimPolicy: Delete
  volumeBindingMode: Immediate
  allowVolumeExpansion: true
  mountOptions:
    - nfsvers=4.2
  ```

  Once created, you can use the StorageClass to create virtual machine images, root volumes, and data volumes.

## References

- [Use Rook Ceph External Storage with Harvester](https://harvesterhci.io/kb/use_rook_ceph_external_storage)
- [Using NetApp Storage on Harvester](https://harvesterhci.io/kb/install_netapp_trident_csi)
- [Third Party Storage Support](https://github.com/harvester/harvester/blob/master/enhancements/20250203-third-party-storage-support.md) 

## Known Issues

> ### 1. infinite image download loop
> 
> Reference issue: https://github.com/harvester/harvester/issues/7993 
> 
> This issue is related to the scratch volume, which is temporarily used to store the image data and is created by CDI. We might see the following error logs on the importer-prime-xxx pod.

Reviewer 3: ```suggestion
### 1. Infinite Image Download Loop

The image download process loops endlessly when the StorageClass for the image uses the LVM CSI driver. This issue is related to the scratch volume, which is created by CDI and is used to temporarily store the image data. When the issue exists in your environment, you might find the following error messages in `importer-prime-xxx` pod logs:

```

Reviewer 3: We usually mention the related issue after the issue description.

    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/pkg/importer/util.go:101
kubevirt.io/containerized-data-importer/pkg/importer.(*HTTPDataSource).Transfer
    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/pkg/importer/http-datasource.go:162
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).initDefaultPhases.func2
    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:173
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessDataWithPause
    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:240
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:149
main.handleImport
    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:188
main.main
    /home/abuild/rpmbuild/BUILD/go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:148
runtime.main
```

> The above log shows that the filesystem created with scratch volume is not enough to store the image data. That's why we see the error `no space left on device`. The scratch volume is created to be the same size as the target volume but adds some filesystem overhead. The filesystem overhead default is 0.055, which means 5.5%. It usually works in most cases. However, if the image size is quite small (usually under 1GB) and its virtual size is very close to the image size, the filesystem overhead of 5.5% might not be enough. To workaround this, we could manually configure the filesystem overhead to a larger value as below:

Reviewer 3: ```suggestion
The message `no space left on device` indicates that the filesystem created using the scratch volume is not enough to store the image data. CDI creates the scratch volume based on the size of the target volume, but some space is lost to filesystem overhead. The default overhead value is `0.055` (equivalent to 5.5%), which is sufficient in most cases. However, if the image size is less than 1 GB and its virtual size is very close to the image size, the default overhead is likely to be insufficient.

The workaround is to increase the filesystem overhead to 20% using the following command:

```

```
# kubectl patch cdi cdi --type=merge -p '{"spec":{"config":{"filesystemOverhead":{"global":"0.2"}}}}'
```

> The above command will set the filesystem overhead to 20%. After that, we should see the image downloading is complete. And don't worry about the higher filesystem overhead. It will not affect the image PVC size. This PVC (scratch volume) will be deleted after the image is imported.
> 
> For more detailed reasons, you can see this [comment](https://github.com/harvester/harvester/issues/7993#issuecomment-2790260841).

Reviewer 3: ```suggestion
The image should be downloaded once the filesystem overhead is increased.

:::note

Increasing the overhead value does not affect the image PVC size. The scratch volume is deleted after the image is imported.

:::

Related issue: [#7993](https://github.com/harvester/harvester/issues/7993) (See this [comment](https://github.com/harvester/harvester/issues/7993#issuecomment-2790260841).)
```
